go/logic/applier.go

```diff
@@ -1225,26 +1234,70 @@ func (this *Applier) ApplyDMLEventQueries(dmlEvents [](*binlog.BinlogDMLEvent))
 		if _, err := tx.Exec(sessionQuery); err != nil {
 			return rollback(err)
 		}
+
+		buildResults := make([]*dmlBuildResult, 0, len(dmlEvents))
 		for _, dmlEvent := range dmlEvents {
 			for _, buildResult := range this.buildDMLEventQuery(dmlEvent) {
 				if buildResult.err != nil {
 					return rollback(buildResult.err)
 				}
-				result, err := tx.Exec(buildResult.query, buildResult.args...)
-				if err != nil {
-					err = fmt.Errorf("%w; query=%s; args=%+v", err, buildResult.query, buildResult.args)
-					return rollback(err)
-				}
 
-				rowsAffected, err := result.RowsAffected()
-				if err != nil {
-					log.Warningf("error getting rows affected from DML event query: %s. i'm going to assume that the DML affected a single row, but this may result in inaccurate statistics", err)
-					rowsAffected = 1
+				buildResults = append(buildResults, buildResult)
+			}
+		}
+
+		execErr := conn.Raw(func(driverConn any) error {
+			ex, ok := driverConn.(driver.ExecerContext)
+			if !ok {
+				return fmt.Errorf("could not cast driverConn to ExecerContext")
+			}
+
+			nvc, ok := driverConn.(driver.NamedValueChecker)
+			if !ok {
+				return fmt.Errorf("could not cast driverConn to NamedValueChecker")
+			}
```

Comment:
I'm wondering if we should just use unsafe casts and panic here. Otherwise, these errors might end up causing retries, but I don't see any sane way how the casts could fail.